providers/dynamodb/shedlock-provider-dynamodb/src/main/java/net/javacrumbs/shedlock/provider/dynamodb/DynamoDBLockProvider.java

```diff
@@ -0,0 +1,238 @@
+        this.table = new DynamoDB(dynamodb).getTable(tableName);
+        this.hostname = Utils.getHostname();
+    }
+
+    @Override
+    public Optional<SimpleLock> lock(LockConfiguration lockConfiguration) {
+        String nowIso = toIsoString(now());
+        String lockUntilIso = toIsoString(lockConfiguration.getLockAtMostUntil());
+
+        UpdateItemSpec request = new UpdateItemSpec()
+                .withPrimaryKey(ID, lockConfiguration.getName())
+                .withUpdateExpression(OBTAIN_LOCK_QUERY)
+                .withConditionExpression(OBTAIN_LOCK_CONDITION)
+                .withValueMap(new ValueMap()
+                        .withString(":lockUntil", lockUntilIso)
+                        .withString(":lockedAt", nowIso)
+                        .withString(":lockedBy", hostname)
+                )
+                .withReturnValues(ReturnValue.UPDATED_NEW);
+
+        try {
+            // There are three possible situations:
+            // 1. The lock document does not exist yet - it is inserted - we have the lock
+            // 2. The lock document exists and lockUtil <= now - it is updated - we have the lock
+            // 3. The lock document exists and lockUtil > now - ConditionalCheckFailedException is thrown
+            UpdateItemOutcome updated = table.updateItem(request);
+            assert lockUntilIso.equals(updated.getItem().getString(LOCK_UNTIL));
+            return Optional.of(new DynamoDBLock(lockConfiguration));
+        } catch (ConditionalCheckFailedException e) {
+            // Condition failed. This means there was a lock with lockUntil > now.
+            return Optional.empty();
+        }
+    }
+
+    private String toIsoString(Instant temporal) {
+        OffsetDateTime utc = temporal.atOffset(ZoneOffset.UTC);
+        return formatter.format(utc);
+    }
+
+    private Instant now() {
+        return Instant.now();
+    }
+
+    private final class DynamoDBLock implements SimpleLock {
+        private final LockConfiguration lockConfiguration;
+
+        private DynamoDBLock(LockConfiguration lockConfiguration) {
+            this.lockConfiguration = lockConfiguration;
+        }
+
+        @Override
+        public void unlock() {
+            // Set lockUntil to now or lockAtLeastUntil whichever is later
+            String unlockTimeIso = toIsoString(lockConfiguration.getUnlockTime());
+            UpdateItemSpec request = new UpdateItemSpec()
+                    .withPrimaryKey(ID, lockConfiguration.getName())
+                    .withUpdateExpression(RELEASE_LOCK_QUERY)
+                    .withValueMap(new ValueMap()
+                            .withString(":lockUntil", unlockTimeIso)
+                    )
+                    .withReturnValues(ReturnValue.UPDATED_NEW);
+            UpdateItemOutcome updated = table.updateItem(request);
+            assert unlockTimeIso.equals(updated.getItem().getString(LOCK_UNTIL));
```

Comment:
I have never seen assert in real code before. Just out of curiosity, how do you use them?

Reply:
To enable assertion checking just add ‚-ea’ to Java commandline. Intelli and Eclipse usually do this automatically when running tests.